question-servers/elements/pl_symbolic_input.py

```diff
@@ -148,6 +146,34 @@ def grade(element_html, element_index, data):
 
     return data
 
+# Safe evaluation of user input to convert from string to sympy expression.
+#
+# Adapted from:
+# https://stackoverflow.com/a/30516254
+#
+# Documentation of ast:
+# https://docs.python.org/3/library/ast.html
+#
+# More documentation of ast:
+# https://greentreesnakes.readthedocs.io/
+
+class Visitor(ast.NodeVisitor):
+    def visit(self, node):
+        if not isinstance(node, self.whitelist):
+            raise ValueError(node)
+        return super().visit(node)
+    whitelist = (ast.Module, ast.Expr, ast.Load, ast.Expression, ast.Call, ast.Name, ast.Num, ast.UnaryOp, ast.UAdd, ast.USub, ast.BinOp, ast.Add, ast.Sub, ast.Mult, ast.Div, ast.Mod, ast.Pow)
```

Comment:
What does `ast.Call` do? Does it allow calling arbitrary functions? I hope not! :-) Things like `ast.Add` seem clearly safe, but I'm slightly nervous about `ast.Module`, `ast.Load`, and `ast.Call`.

Reply:
Good questions.

Just so we are on the same page, `ast` does not produce a python code object (it does not call anything). It produces a tree of objects (`ast.Call` is an example of an object) that can be compiled into a python code object (and evaluated).

`ast.Call` is the object that allows calling a function. Yes, it allows calling arbitrary functions. There aren't special `ast` objects to contain special functions - it's an all or nothing sort of thing.

The `ast.Call` object is converted to a python code object using `compile`, and is actually called used `eval`, which takes two optional arguments - `globals` and `locals`. These are dictionaries that describe the global and local namespace that will be used when evaluating the python code object.

From [the eval() docs](https://docs.python.org/3/library/functions.html#eval):
> If the globals dictionary is present and lacks ‘__builtins__’, the current globals are copied into globals before expression is parsed.

This is why the code specifies `globals={'__builtins__': None}`. The global namespace is empty.

Also from [the eval() docs](https://docs.python.org/3/library/functions.html#eval):
> If the locals dictionary is omitted it defaults to the globals dictionary.

This is why the code specifies a locals dictionary. In particular, it's this:
```python
{'cos': sympy.cos, 'sin': sympy.sin, 'tan': sympy.tan, 'exp': sympy.exp, 'log': sympy.log, 'sqrt': sympy.sqrt, 'pi': sympy.pi}
```

Keys specify function names (or other things in the local namespace during `eval`) and values specify what these functions actually are - I've mapped them all to `sympy` function objects. If a function name does not appear as a key in this dictionary, then it cannot be evaluated (because with respect to the global and local namespaces, it simply does not exist) - `eval` will throw an exception.

The bottom line is that `ast.Call` allows arbitrary functions, but `eval` applied to a compiled `ast.Call` object - with local and global namespaces specified - does not.

`ast.Module` is simply the root of the `ast` object tree ([see the docs](https://greentreesnakes.readthedocs.io/en/latest/tofrom.html)).

`ast.Load` is similar to `ast.Call` but allows the use of variables - it is the `ast` object that loads the value of a variable. If that variable exists in the global or local namespace, then it gets assigned during `eval` - if not, an exception is raised. You'll note `pi` in the list of locals - this is an example of a variable that will get assigned by evaluating a compiled `ast.Load` object. Similarly, you'll note that I add a key/value pair for each variable students are allowed to use in their symbolic expression (if any) - so, `'x': sympy.Symbol('x')`, for example. Use of any other variable will result in an error.

CPU and memory resource vulnerabilities are still present, although these are largely addressed by the fact that PL has timeouts.

Other vulnerabilities such as those discussed [here](https://nedbatchelder.com/blog/201206/eval_really_is_dangerous.html) are not, I think, an issue for us. I do not pretend to be an expert in computer security though.